### PR TITLE
Add Calendar Arithmetic

### LINKED ISSUE #1470: Write common framework for calendar arithmetic
Part of https://github.com/unicode-org/icu4x/issues/534

Currently we reimplement the `offset_date()` and `until()` `Calendar` functions for each calendar, with Gregorian proxying to ISO. Most calendars are quite similar here:

 - Gregorian and Julian just differ by _when_ the leap years occur
 - Indian National differs by the exact month sizing
 - Coptic/Ethiopian occasionally have an extra epagomenal month in the year.

All of this can be analyzed as calendars having different month sizing.

It would be ideal if this arithmetic did not need to be reimplemented each time.

Furthermore, dealing with the ISODay/etc types is kind of annoying; it's nice for the types to be typesafe for the user but there's plenty of conversion arithmetic/etc going on.

First, I propose we have a single internal type for all solar dates. It can later be expanded to handle lunisolar and lunar calendars.

```rust
struct ArithmeticDate<C: CalendarArithmetic> {
    year: i32,
    month: u8,
    day: u8,
}

struct ISODateInner(ArithmeticDate<ISO>);
struct JulianDateInner(ArithmeticDate<Julian>);
// etc
```

We also have an internal trait `CalendarArithmetic` that works the following way:

```rust
trait CalendarArithmetic {
   fn month_lengths(year: i32) -> [u8; 12]; // later, &[u8]
}
```

Finally, `ArithmeticDate` has implementations of `offset_date()`, `until()`, `days_in_year()`, `days_in_month()` that rely on the month length information to perform calculations. Calendars using it (currently: ISO and Julian, since Buddhist and Gregorian just use ISO) can just proxy to it.

New calendars that fit this model will be pretty easy to implement: you still have to implement the conversion function to/from ISO (if we want we can add `foo_from_fixed()`/etc to `CalendarArithmetic` to consolidate further), but after that you can just implement `month_lengths()` and do a simple wrapper.

This will make it really easy to implement the Indian national calendar. Coptic and Ethiopian would need an extra short month. Islamic will have a more complex implementation of `month_lengths()` but it'll be much the same.

Hebrew _potentially_ could be handled by allowing for one of the `month_lengths()` to be zero, there are just a bunch of tricky bits in writing `offset()` and `until()` functions that can work with that.

### PR DESCRIPTION
**Pull request by pandusonu2**
Fixes #1470

### CODE REVIEW

File: components/calendar/src/calendar_arithmetic.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+pub trait CalendarArithmetic: Calendar {
+    fn month_lengths(year: i32) -> [u8; 12];
+    fn months_for_every_year() -> Option<u8>;
+    fn is_leap_year(year: i32) -> bool;
+}
+
+impl<C: CalendarArithmetic> ArithmeticDate<C> {
+    #[inline]
+    pub fn offset_date(&self, mut offset: DateDuration<C::Calendar>) {
```
Comment (Manishearth) [inline components/calendar/src/calendar_arithmetic.rs:23]:
```suggestion
    pub fn offset_date(&self, mut offset: DateDuration<C>) {
```

**Review by Manishearth — CHANGES_REQUESTED** [review]

File: components/calendar/src/julian.rs [hunk 2]
```diff
@@ -38,23 +59,19 @@ impl Calendar for Julian {
     }
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:59]: nit: update this to use calendararithmetic

File: components/calendar/src/calendar_arithmetic.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::{Calendar, DateDuration, DateDurationUnit};
+use core::marker::PhantomData;
+
+pub struct ArithmeticDate<C: CalendarArithmetic> {
```
Comment (Manishearth) [inline components/calendar/src/calendar_arithmetic.rs:8]: needs derive(Copy, Clone)

File: components/calendar/src/julian.rs [hunk 4]
```diff
@@ -65,22 +82,17 @@ impl Calendar for Julian {
...
     /// The calendar-specific year represented by `date`
     fn year(&self, date: &Self::DateInner) -> types::Year {
-        date.0.year.into()
+        IsoYear(date.0.year).into()
     }
 
     /// The calendar-specific month represented by `date`
     fn month(&self, date: &Self::DateInner) -> types::Month {
-        date.0.month.into()
+        IsoMonth(date.0.month).into()
```
Comment (pandusonu2) [inline components/calendar/src/julian.rs:95]: This seems to be wrong, as `IsoMonth(u8)` is not public. Also, can't just return `date.0.month` as `u8` does not impl `for types::Month`

What is the suggested approach here?
Delete all usage of IsoMonth entirely and `Impl types::Month` for u8?

cc: @Manishearth

Comment (Manishearth) [reply components/calendar/src/julian.rs:95]: > and `Impl types::Month` for u8?

This won't work, it's not a trait. Call `date.0.month.try_into().unwrap()` here.

Comment (Manishearth) [reply components/calendar/src/julian.rs:95]: Actually, `types::Month` is a different type, you shouldn't need to be using IsoMonth here

https://unicode-org.github.io/icu4x-docs/doc/icu/calendar/types/struct.Month.html

Comment (Manishearth) [reply components/calendar/src/julian.rs:95]: also for `day_of_month()` below it should be `types::DayOfMonth(date.0.day)`, probably.

Perhaps add a method to `ArithmeticDate` that does this for convenience

File: components/calendar/src/julian.rs [hunk 5]
```diff
@@ -65,37 +84,35 @@ impl Calendar for Julian {
...
-            0,
-            u8::from(date1.0.day) as i32 - u8::from(date2.0.day) as i32,
-        )
+        date1.0.until(date2.0, _largest_unit, _smallest_unit)
     }
 
     /// The calendar-specific year represented by `date`
     fn year(&self, date: &Self::DateInner) -> types::Year {
-        date.0.year.into()
+        crate::gregorian::year_as_gregorian(date.0.year)
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:92]: nit: drop in a comment saying that the Julian year has the same era scheme as Gregorian

**Review by Manishearth — APPROVED** [review]
This looks great! One minor nit and we can land this.

File: components/calendar/src/julian.rs (hunk 5 above)
Comment (pandusonu2) [reply components/calendar/src/julian.rs:92]: Done

**Review by Manishearth — APPROVED** [review]

**Pull request merged**